Runtimes/Core/cmake/modules/EmitSwiftInterface.cmake

```diff
@@ -2,32 +2,53 @@
 
 # TODO: CMake should learn how to model library evolution and generate this
 #       stuff automatically.
-
+if(POLICY CMP0195)
+  message(AUTHOR_WARNING [=[
+    This version of CMake supports the Swift module directory structure
+    Please consider checking the implementation of `emit_swift_interface``
+    to ensure it behaves as expected
+    ]=])
+endif()
 
 # Generate a swift interface file for the target if library evolution is enabled
 function(emit_swift_interface target)
   # Generate the target-variant binary swift module when performing zippered
   # build
+  # Clean this up once CMake has nested swiftmodules in the build directory:
+  # https://gitlab.kitware.com/cmake/cmake/-/merge_requests/10664
+  # https://cmake.org/cmake/help/git-stage/policy/CMP0195.html
+
+  # We can't expand the Swift_MODULE_NAME target property in a generator
+  # expression or it will fail saying that the target doesn't exist.
+  get_target_property(module_name ${target} Swift_MODULE_NAME)
+  if(NOT module_name)
+    set(module_name ${target})
+  endif()
+  file(MAKE_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/${module_name}.swiftmodule")
+  target_compile_options(${target} PRIVATE
+    "$<$<COMPILE_LANGUAGE:Swift>:SHELL:-emit-module-path ${CMAKE_CURRENT_BINARY_DIR}/${module_name}.swiftmodule/${SwiftCore_MODULE_TRIPLE}.swiftmodule>")
   if(SwiftCore_VARIANT_MODULE_TRIPLE)
-    set(variant_module_tmp_dir "${CMAKE_CURRENT_BINARY_DIR}/${target}-${SwiftCore_VARIANT_MODULE_TRIPLE}")
-    file(MAKE_DIRECTORY "${variant_module_tmp_dir}")
     target_compile_options(${target} PRIVATE
-      "$<$<COMPILE_LANGUAGE:Swift>:SHELL:-emit-variant-module-path ${variant_module_tmp_dir}/${target}.swiftmodule>")
+      "$<$<COMPILE_LANGUAGE:Swift>:SHELL:-emit-variant-module-path ${CMAKE_CURRENT_BINARY_DIR}/${module_name}.swiftmodule/${SwiftCore_VARIANT_MODULE_TRIPLE}.swiftmodule>")
   endif()
+  add_custom_command(OUTPUT "${CMAKE_CURRENT_BINARY_DIR}/${module_name}.swiftmodule/${SwiftCore_MODULE_TRIPLE}.swiftmodule"
+    DEPENDS ${target})
+  target_sources(${target}
+    INTERFACE
+      $<BUILD_INTERFACE:${CMAKE_CURRENT_BINARY_DIR}/${module_name}.swiftmodule/${SwiftCore_MODULE_TRIPLE}.swiftmodule>)
 
   # Generate textual swift interfaces is library-evolution is enabled
   if(SwiftCore_ENABLE_LIBRARY_EVOLUTION)
```

Comment:
I took all the feedback that @compnerd left in #80881, except for the one asking to remove the check for `SwiftCore_ENABLE_LIBRARY_EVOLUTION` -- both because I don't fully appreciate the background about it, and because this PR has already a lot of refactoring.